### PR TITLE
[python-package] Add `feature_names_in_` attribute for scikit-learn estimators (fixes #6279)

### LINKED ISSUE #6279: [python-package] Support `feature_names_in_` attribute via sklearn API
## Summary

sklearn API supports `feature_names_in_` attribute on a fitted model ([SLEP007](https://scikit-learn-enhancement-proposals.readthedocs.io/en/latest/slep007/proposal.html)), which remembers the feature names/columns that went into the `model.fit` method. This can be very useful information, and is a standard worth conforming to. Afaiu right now that information is available in the booster:

```py
est.booster_.feature_name()
```

It shouldn't be too hard to conform to also expose that information via `feature_names_in_` attribute 🙏 

## Motivation

It would conform to the sklearn API standards, improve usability of LightGBM models, especially when used along with other sklearn models and Pipelines.

## References

* https://scikit-learn-enhancement-proposals.readthedocs.io/en/latest/slep007/proposal.html

### PR DESCRIPTION
**Pull request by nicklamiller**
Fixes: #6279

Related: https://github.com/scikit-learn/scikit-learn/issues/28337

### CODE REVIEW

File: python-package/lightgbm/sklearn.py
Comment (jameslamb) [inline python-package/lightgbm/sklearn.py:1073]: Please update this with the following:

* remove "sklearn-style property for" and instead just say what it is, something like "names for features"
* this should only be available in a fitted model, right? If so, please guard it like this:

https://github.com/microsoft/LightGBM/blob/cc733f8595267f313886f92ed5d1285010ba8f3f/python-package/lightgbm/sklearn.py#L993-L994

* explain in the docs what will happen when accessing this attribute if you never provided feature names (e.g. just passed raw `numpy` arrays as training data)

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for this!

But please add some unit tests in https://github.com/microsoft/LightGBM/blob/master/tests/python_package_test/test_sklearn.py covering the following:

* what happens when you try to access that attribute on an unfitted estimator
* that that attribute returns the exact expected values in the following situations:
   - trained with feature names (in each of the ways feature names can be provided, e.g. do you get them automatically using `pandas` as input?)
   - trained without feature names

**Comment by jameslamb** [comment]
In https://github.com/scikit-learn/scikit-learn/issues/28337#issuecomment-1921071816, I noticed someone said

> this feature comes for free if you inherit from `BaseEstimator`

`lightgbm`'s scikit-learn estimators do inherit from `BaseEstimator`

https://github.com/microsoft/LightGBM/blob/cc733f8595267f313886f92ed5d1285010ba8f3f/python-package/lightgbm/sklearn.py#L1072

https://github.com/microsoft/LightGBM/blob/cc733f8595267f313886f92ed5d1285010ba8f3f/python-package/lightgbm/sklearn.py#L430

https://github.com/microsoft/LightGBM/blob/cc733f8595267f313886f92ed5d1285010ba8f3f/python-package/lightgbm/sklearn.py#L15-L17

https://github.com/microsoft/LightGBM/blob/cc733f8595267f313886f92ed5d1285010ba8f3f/python-package/lightgbm/compat.py#L106

https://github.com/microsoft/LightGBM/blob/cc733f8595267f313886f92ed5d1285010ba8f3f/python-package/lightgbm/compat.py#L83

If you get into this and find that `lightgbm` is actually getting that attribute via inheriting from `BaseEstimator`, don't give up on the PR! Those tests I mentioned would still be very valuable to catch changes to that support in the future and to be sure that `lightgbm`'s integration with it has the expected behavior.

**Comment by nicklamiller** [comment]
@jameslamb Thank you for the great feedback! I'm working on adding these suggestions in.

Is there a way you recommend recreating the development environment? I couldn't find info on this in the `CONTRIBUTING.md` so started to mimic the logic specified in [.ci/test.sh](https://github.com/microsoft/LightGBM/blob/master/.ci/test.sh) but having to specify different global variables prevents this from being a quick execution/way to setup the environment. Just want to make sure I'm not missing a quicker way.

Thanks in advance!

**Comment by jameslamb** [comment]
Thanks! There isn't a well-documented way to set up a local development environment for the Python package today, it's something I'd like to add soon.

Here's how I develop on LightGBM:

1. Create a `conda` environment ([using miminiforge](https://github.com/conda-forge/miniforge))

```shell
mamba create \
    --name lgb-dev \
    cloudpickle \
    dask \
    distributed \ 
    joblib \
    matplotlib \
    numpy \
    python-graphviz \
    pytest \
    pytest-cov \
    python=3.11 \
    scikit-learn \
    scipy
```

2. build the C++ library one time (assuming you're making Python-only changes)

```shell
rm -rf ./build
mkdir ./build
cd ./build
cmake ..
make -j4 _lightgbm
```
3. make changes to the Python code
4. install the Python package in the conda environment

```shell
source activate lgb-dev
sh build-python.sh install --precompile
```

5. run the tests

```shell
pytest testss/python_package_test
```

6. repeat steps 3-5 until you're confident in your changes
7. run the auto-formatting and some of the linting stuff (this is a work in progress, see #6304)

```shell
pre-commit run --all-files
```

**Comment by nicklamiller** [comment]
> If you get into this and find that `lightgbm` is actually getting that attribute via inheriting from `BaseEstimator`, don't give up on the PR!

It turns out sklearn only adds this attribute if the input data has feature names, while LightGBM adds this attribute and column names of the format "Column_{i}". I've added a comment to a test to highlight this difference with sklearn.

**Comment by nicklamiller** [comment]
@microsoft-github-policy-service agree

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for this!

But this does not look like it's meeting the expectations described in https://scikit-learn-enhancement-proposals.readthedocs.io/en/latest/slep007/proposal.html.

I re-read that tonight, and saw the following

> **Input Feature Names**
>
> *The input feature names are stored in a fitted estimator in a `feature_names_in_` attribute, and are taken from the given input data, for instance a `pandas` data frame. 
> *This attribute will be `None` if the input provides no feature names. The `feature_names_in_` attribute is a 1d NumPy array with object dtype and all elements in the array are strings.*

> **Output Feature Names**
> *A fitted estimator exposes the output feature names through the `get_feature_names_out` method. The output of `get_feature_names_out` is a 1d NumPy array with object dtype and all elements in the array are strings. Here we discuss more in detail how these feature names are generated. Since for most estimators there are multiple ways to generate feature names, this SLEP does not intend to define how exactly feature names are generated for all of them. It is instead a guideline on how they could generally be generated.*

So I think the following needs to be done:

* `feature_names_in_` should return a 1D `numpy` array, not a list
* `get_feature_names_out()` function should be implemented (right? or is that only for estimators that define `.transform()`?)

---

There is also still something that's really bothering me about this in general, that I think we need to get a clear answer on before going further.

This comment claims that you get these things for free if you inherit from `BaseEstimator`: https://github.com/scikit-learn/scikit-learn/issues/28337#issuecomment-1921071816

But `lightgbm.sklearn.LGBMModel` and everything inheriting from it do inherit from `BaseEstimator`. I've asked about this here: https://github.com/scikit-learn/scikit-learn/issues/28337#issuecomment-2026628918.

Up to you if you'd like to wait for `scikit-learn` maintainers to respond there before working on the other things I've requested here.

File: python-package/lightgbm/sklearn.py
Comment (jameslamb) [reply python-package/lightgbm/sklearn.py:1128]: Actually, let's wait on this until we get more clarity on https://github.com/scikit-learn/scikit-learn/issues/28337#issuecomment-2026628918.

**Comment by nicklamiller** [comment]
@jameslamb given that `_validate_data` needs to be called in order to get these attributes for free from `BaseEstimator`, would it make sense to call this method within the LGBM estimators' `fit` methods (like many other sklearn estimators, one example: https://github.com/scikit-learn/scikit-learn/issues/27907#issuecomment-1848837382)?

One different behavior between LGBM and sklearn is that LGBM assigns artificial names to features if the features are unnamed, whereas sklearn doesn't create artificial names, and also doesn't create the `feature_names_in_` attribute. So for numpy arrays, even calling `_validate_data` within `fit` wouldn't make this attribute accessible.

I wanted to confirm that we want add `_validate_data`, but to also keep the behavior of setting names when they're not present.

**Comment by nicklamiller** [comment]
>     feature_names_in_ should return a 1D numpy array, not a list
Sounds good, will fix.

>     get_feature_names_out() function should be implemented (right? or is that only for estimators that define .transform()?)
I have less of an opinion on this one, but based on the SLEP:

> Scope[](https://scikit-learn-enhancement-proposals.readthedocs.io/en/latest/slep007/proposal.html#scope)
> 
> The API for input and output feature names includes a feature_names_in_ attribute for all estimators, and a get_feature_names_out method for any estimator with a transform method, i.e. they expose the generated feature names via the get_feature_names_out method.

it does look like it should be specifically for estimators with the `transform` method.

**Comment by jameslamb** [comment]
> I wanted to confirm that we want to add `_validate_data`

That method being prefixed with a `_` suggests to me that it's an internal implementation detail of `scikit-learn` that could be changed in a future release of that library.

Can you find me some authoritative source saying that projects implementing their own estimators are encouraged to call that method? The comment you linked above is a specific recommendation from a `scikit-learn` maintainer about what to do for 2 estimators **within `scikit-learn`**... I don't interpret that as encouragement that other projects should call it.

`xgboost` does not: https://github.com/search?q=repo%3Admlc%2Fxgboost%20%22_validate_data%22&type=code

but `catboost` does: https://github.com/catboost/catboost/blob/19b60a20b2b1733c528b40c6c9ebe2f3d1f5dbde/contrib/python/scikit-learn/py3/sklearn/base.py#L537

Let's please pause on this work until some `scikit-learn` maintainer gives an authoritative answer on https://github.com/scikit-learn/scikit-learn/issues/28337.

**Comment by jameslamb** [comment]
Alright @nicklamiller I've read through the response at https://github.com/scikit-learn/scikit-learn/issues/28337#issuecomment-2068815197 carefully... I think we have enough information to proceed.

I think we should do the following here in LightGBM:

- add `feature_names_in_`
- add `get_feature_names_out`()
- **do not** directly call `_validate_data()` or any other `scikit-learn` methods prefixed with `_`

Let's try to as closely as possible follow the way that `xgboost` has handled this in their Python package.

Are you interested in continuing this?

**Comment by nicklamiller** [comment]
@jameslamb I tried to closely follow xgboost w.r.t `feature_names_in_`, which is similar in [its implementation](https://github.com/dmlc/xgboost/blob/d2d01d977a5ba07e55ed9230e2bd83539e81d184/python-package/xgboost/sklearn.py#L1291-L1302) in that feature names are obtained from another attribute from a `_Booster` object which is only returned if the model is fitted. If feature names can’t be accessed because the model is not fitted, a `NotFittedError` is thrown (so similar to the `LGBMNotFittedError` that we throw here).

`xgboost` doesn’t implement the `get_feature_names_out` method, so I followed `scikit-learn`’s approach for this. Most objects that have this method are subclasses of `TransformerMixin` some examples:

- [`AdditiveChi2Sampler`](https://github.com/scikit-learn/scikit-learn/blob/7398114752a056edffb6c768393e61c32d56a6c6/sklearn/kernel_approximation.py#L738-L764)
- [`RandomTreesEmbedding`](https://github.com/scikit-learn/scikit-learn/blob/7398114752a056edffb6c768393e61c32d56a6c6/sklearn/ensemble/_forest.py#L2933-L2963)

The few classes that inherit from an estimator-like mixin (`ClassifierMixin` and `RegressorMixin`) appear to be meta-estimators that specify features for each individual estimator that the meta estimator is comprised of:

- [`VotingClassifier`](https://github.com/scikit-learn/scikit-learn/blob/7398114752a056edffb6c768393e61c32d56a6c6/sklearn/ensemble/_voting.py#L517-L552)
- [`VotingRegressor`](https://github.com/scikit-learn/scikit-learn/blob/7398114752a056edffb6c768393e61c32d56a6c6/sklearn/ensemble/_voting.py#L734-L753)


So it looks like `get_feature_names_out` is in fact defined almost exclusively for transformers. I’ve still added it here but since `LGBMModel` and its subclasses don’t apply transformations to features and aren’t meta-estimators, I simply return `feature_names_in_` within the method.

Please let me know how this looks!

File: tests/python_package_test/test_sklearn.py
Comment (jameslamb) [inline tests/python_package_test/test_sklearn.py:1293]: There is quite a lot of repetition in these tests. Generally I'm supportive of repetition in tests in favor of making it easier to diagnose issues or selectively include / exclude tests... but for the cases of `.feature_names_in_` and `.get_feature_names_out()`, I think the tests should be combined.

So can you please reduce these 4 tests down to 2? One for `numpy` input without feature names, one for `pandas` input with feature names?

Ending with assertions like this:

```python
expected_col_names = np.array([f"Column_{i}" for i in range(X.shape[1])]
np.testing.assert_array_equal(model.feature_names_in_, expected_col_names)
np.testing.assert_array_equal(model.get_feature_names_out(), expected_col_names)
```

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for much for your persistence with this!

The implementations look good to me. I just have some recommendations on the tests.

> *So it looks like `get_feature_names_out` is in fact defined almost exclusively for transformers. I’ve still added it here but since `LGBMModel` and its subclasses don’t apply transformations to features and aren’t meta-estimators, I simply return `feature_names_in_` within the method.*

I asked for it to be added based on this statement form @adrinjalali in https://github.com/scikit-learn/scikit-learn/issues/28337#issuecomment-2068815197: *"scikit-learn estimators all add ... `get_feature_names_out()`"*.

I think the choice to just return `feature_names_in_` makes sense for `lightgbm`, thank you.

---

Please note... I am going to try to push this week to get LightGBM v4.4.0 out, so we can have a release up before NumPy 2.0 is released June 16th (https://github.com/microsoft/LightGBM/pull/6439#issuecomment-2133974437). So if you don't have time to address this feedback in the next couple days, then this will be in the **next** release of LightGBM after that.

**Comment by jameslamb** [comment]
Alright, thanks for that clarification.

@nicklamiller please remove that method then.

**Comment by nicklamiller** [comment]
@jameslamb Thanks very much for the suggestions! I removed `get_feature_names_out()` and fixed up the tests as specified above, please let me know how this looks.

**Comment by jameslamb** [comment]
Thanks very much for keeping this up to date with `master`.

I'll review again soon, but it's too late for this to make it into the v4.4.0 release (#6439). Sorry about that, but we are rushing to get that release up before `numpy` 2.0 comes out 2 days from now.

**Review by jameslamb — APPROVED** [review]
Looks great!

Thank you VERY much for your patience and persistence on this. We really appreciate it and would love to have you come back and contribute more in the future!

**Comment by jameslamb** [comment]
@borchero could you re-run the failing Azure check and merge this if it passes? I'll be away from my laptop for the next day or 2

**Comment by jameslamb** [comment]
I just restarted that failing CI job, will check back in a bit.

**Comment by borchero** [comment]
Sorry @jameslamb, I saw your comment only now! Unfortunately, I also can't restart Azure jobs, my login continues to fail 🙄

**Comment by jameslamb** [comment]
> *Unfortunately, I also can't restart Azure jobs, my login continues to fail 🙄*

No problem! I know, I also sometimes have to log into Azure DevOps multiple times and in different ways to finally get in. I'll send you a private message with some tips to try.

**Comment by jameslamb** [comment]
Thanks so much @nicklamiller ! I'm sorry this was such a long process for you. We hope you'll consider coming back and contributing more in the future.

If you're interested in helping with the Python package specifically, this would be a good next place to go: #6361

**Comment by nicklamiller** [comment]
@jameslamb thank you for all the help and guidance, I'm looking forward to contributing more!

**Pull request merged**
